dev/Interop/StoragePickers/FileSavePicker.cpp

```diff
@@ -0,0 +1,176 @@
+        parameters.FileTypeFilterPara = PickerCommon::CaptureFilterSpec(parameters.FileTypeFilterData, m_fileTypeChoices.GetView());
+
+    }
+
+    winrt::Windows::Foundation::IAsyncOperation<winrt::Microsoft::Windows::Storage::Pickers::PickFileResult> FileSavePicker::PickSaveFileAsync()
+    {
+        auto logTelemetry{ StoragePickersTelemetry::FileSavePickerPickSingleFile::Start(m_telemetryHelper) };
+
+        PickerCommon::PickerParameters parameters{};
+        CaptureParameters(parameters);
+        auto defaultFileExtension = m_defaultFileExtension;
+        auto suggestedSaveFile = m_suggestedSaveFile;
+        auto suggestedFileName = m_suggestedFileName;
+        auto fileTypeChoices = m_fileTypeChoices;
+
+        co_await winrt::resume_background();
+        auto cancellationToken = co_await winrt::get_cancellation_token();
+        if (cancellationToken())
+        {
+            logTelemetry.Stop(m_telemetryHelper, false);
+            co_return nullptr;
+        }
+
+        auto dialog = create_instance<IFileSaveDialog>(CLSID_FileSaveDialog, CONTEXT_ALL);
+        parameters.ConfigureDialog(dialog);
+
+        if (!PickerCommon::IsHStringNullOrEmpty(defaultFileExtension))
+        {
+            check_hresult(dialog->SetDefaultExtension(defaultFileExtension.c_str()));
+        }
+
+        if (!PickerCommon::IsHStringNullOrEmpty(suggestedFileName))
+        {
+            check_hresult(dialog->SetFileName(suggestedFileName.c_str()));
+        }
+
+        if (suggestedSaveFile != nullptr)
+        {
+            winrt::com_ptr<IShellItem> shellItem;
+            check_hresult(SHCreateItemFromParsingName(suggestedSaveFile.Path().c_str(), nullptr, IID_PPV_ARGS(shellItem.put())));
+            check_hresult(dialog->SetSaveAsItem(shellItem.get()));
+        }
+
+        {
+            auto hr = dialog->Show(parameters.HWnd);
+            if (FAILED(hr))
+            {
+                logTelemetry.Stop(m_telemetryHelper, false);
+                co_return nullptr;
+            }
+        }
+
+        winrt::com_ptr<IShellItem> shellItem{};
+        check_hresult(dialog->GetResult(shellItem.put()));
+
+        // Get the file path from the dialog
+        wil::unique_cotaskmem_string filePath{};
+        check_hresult(shellItem->GetDisplayName(SIGDN_FILESYSPATH, filePath.put()));
+        winrt::hstring pathStr(filePath.get());
+
+        wil::unique_cotaskmem_string fileName;
+        check_hresult(shellItem->GetDisplayName(SIGDN_NORMALDISPLAY, fileName.put()));
+        std::wstring fileNameStr(fileName.get());
+
+        // Check if the file name has an extension
+        if (fileNameStr.find_last_of(L".") == std::wstring::npos)
+        {
+            // If the user defined file name doesn't have an extension,
+            //     automatically use the first extension from the first category in fileTypeChoices.
+            auto firstCategory = fileTypeChoices.First().Current();
```

Comment:
>fileTypeChoices.First() [](http://example.com/codeflow?start=33&length=23)

Is it legal to not pass in any FileTypeChoices? I think we decided it is, in which case some handling is required here.

Reply:
Thank you very much for your for catching this mistake, a guard for empty fileTypeChoices is necessary here.
For file save pickers, filter behavior might be different for open ones. for open picker we've decide to use wildcard fallback when calling pickers with empty filters. But for save pickers, since we are creating a new file, a concrete extension is required, thus we can not fall back to use wildcard files.